torch/distributed/pipelining/schedules.py

```diff
@@ -2161,3 +2168,148 @@ def get_schedule_class(schedule_name: str):
             f"Unknown schedule name '{schedule_name}'. The valid options are {list(schedule_map.keys())}"
         )
     return schedule_map[lowercase_keys[lowercase_schedule_name]]
+
+
+def _simulate_comms_compute(
```

Comment:
comment on function's purpose? Do i have the correct understanding that it is intended to be used after adding comm ops in `_add_send_recv` to reorder the communication ops?

Reply:
ill add a comment. 

No, the purpose is not to reorder ops.  In fact it should never reorder ops, but it may add bubbles.  The 2 points for the function are (1) to determine if the schedule is expected to deadlock, (2) and assuming it runs without deadlock, simulate the logical ordering of ops per timestep.  I don't require users to put 'bubbles' into their schedule definitions, but the simulator would add the bubbles in based on whether there are timesteps where no action is 'ready' to execute on the local rank given its comm dependencies.